src/lm/static_probe_map.cpp

```diff
@@ -0,0 +1,70 @@
+/**
+ * @file static_probe_map.cpp
+ * @author Sean Massung
+ */
+
+#include <cstring>
+#include "lm/static_probe_map.h"
+
+namespace meta
+{
+namespace lm
+{
+static_probe_map::static_probe_map(const std::string& filename,
+                                   uint64_t num_elems)
+    : table_{filename, static_cast<uint64_t>((num_elems / 0.7) * 2)},
+      hash_{seed_}
+// load factor of 0.7; x2 for keys and vals
+{
+}
+
+void static_probe_map::insert(const std::string& key, float prob, float backoff)
+{
+    auto hashed = hash(key);
+    auto idx = (hashed % (table_.size() / 2)) * 2;
+
+    while (true)
+    {
+        if (table_[idx] == uint64_t{0})
+        {
+            table_[idx] = hashed;
+
+            // pack prob and float into uint64_t slot next to key val
+            uint64_t buf = 0;
+            std::memcpy(&table_[idx + 1], &prob, sizeof(float));
+            std::memcpy(&buf, &backoff, sizeof(float));
+            table_[idx + 1] |= (buf << 32);
```

Comment:
Wait, is this doing the right thing? Is this assuming a specific endianness?

Below is a program that (I think) is testing how you're doing this. It's got another function that's written how I would have done it.

```cpp
#include <iostream>
#include <vector>
#include <cstdint>
#include <cstring>

void seanbits()
{
    uint64_t bytes;
    float prob = 0.2;
    float backoff = 0.5;

    std::memcpy(&bytes, &prob, sizeof(float));
    uint64_t buf;
    std::memcpy(&buf, &backoff, sizeof(float));
    bytes |= (buf << 32);

    uint32_t outbuf = (bytes >> 32);
    float probout;
    float backoffout;
    std::memcpy(&probout, &bytes, sizeof(float));
    std::memcpy(&backoff, &outbuf, sizeof(float));

    std::cout << "prob: " << prob << ", out: " << probout << "\n"
              << "back: " << backoff << ", out: " << backoffout << std::endl;
}

void chasebits()
{
    uint64_t bytes;
    float prob = 0.2;
    float backoff = 0.5;

    char* buf = reinterpret_cast<char*>(&bytes);
    std::memcpy(buf, &prob, sizeof(float));
    std::memcpy(buf + sizeof(float), &backoff, sizeof(float));

    float probout;
    float backoffout;
    std::memcpy(&probout, buf, sizeof(float));
    std::memcpy(&backoff, buf + sizeof(float), sizeof(float));

    std::cout << "prob: " << prob << ", out: " << probout << "\n"
              << "back: " << backoff << ", out: " << backoffout << std::endl;
}

int main()
{
    seanbits();
    chasebits();
}
```

Output:
```
prob: 0.2, out: 0.2
back: 0.5, out: 0
prob: 0.2, out: 0.2
back: 0.5, out: 0.5
```

Now, I'm not for sure if the error is in reading or writing, but I think the way I'm writing it is more clear about the intent (and should(?) work regardless of endianness, provided that the model files are generated and read on the same machine).

In any case, this logic should probably move to `lm_node`, which has a `write_packed()` or something so that this logic is all in one place and can be easily unit tested.